### ISSUE TITLE
The _substituteVars function throws an illegal key type for substitution map if it encounters and array while mapping the variables in custom-environment-variables.json

### ISSUE THREAD

**Issue by vchouhan** [lorenwest/node-config#472]
**Note: for support questions, please use [StackOverflow](https://stackoverflow.com/questions/tagged/node-config) and tag your question with `node-config`**. *This repository's issues are reserved for feature requests and bug reports.*

Before submitting a bug report, please search the issue tracker the wiki first. Many issues have already been discussed.

The wiki is located at: https://github.com/lorenwest/node-config/wiki

### I'm submitting a ... 
  - [ x ] bug report

### What is the current behavior?
The _substituteVars function throws an illegal key type for substitution map if it encounters and array while mapping the variables in the `custom-environment-variables.json`
*Create `custom-environment-variables.json` file and add a properties to pull values from env file:

"services": {
                "messagehub": [
                    {
                        "label": "messagehub",
                        "credentials": {
                            "api_key": "API_KEY", // from env 
                            "kafka_rest_url": "REST_URL" // from env
                        }
                    }
                ]
            }
`
<img width="1189" alt="screen shot 2018-02-21 at 07 58 13" src="https://user-images.githubusercontent.com/1222739/36481182-0a38dcc6-16dd-11e8-9455-1fab5b2c744d.png">

### What is the expected behavior?
The _substituteVars function in config.js throws an error. `Illegal key type for substitution map at components.services: object` as the function is only validating 'strings' and 'objects' and since I've an 
### Please tell us about your environment:

 -  node-config version: 1.29.4
 -  node-version: 9.0
 
### Other information
 - No, other environment defined.
(e.g. are you using an environment besides Node.js?)

**Comment by markstos** [comment]
@lorenwest should we even try to support arrays in `custom-environment-variables.json`? It seems tricky since one config can override an array of a previous config, changing the location of particular values in the array. Clearly it hasn't work before, so officially not supporting it would be backwards compatible.  Workarounds include:

  * Referencing environment variables directly from an array
  * Using the JavaScript file type and a `defer()` function to reference another config value from an array, which in turn has a `custom-environment-variables.json` mapping.

**Comment by lorenwest** [comment]
If a configuration value is an array, I support replacing the entire array but no deeper. I'm not sure based on the description if the attempt is to replace the entire array, or an element of the array. If the former, then I'd say it's worth consideration as a bug. If the latter, then we wouldn't consider it a bug if it doesn't work.  Too many issues with replacing array values.

**Comment by markstos** [comment]
He's trying to express "for the configuration key "messagehub", the first value of that array should be set to the environment variable named "API_KEY", using `custom-environment-variables.json`. I'm just not sure we can reliably get the environment variable substitution to work inside an array.

**Comment by lorenwest** [comment]
That's an example of the latter case and I don't feel array element replacement is something we want to support. Arrays in configs are fine, but like strings, they're replaced all or nothing. No sub-string replacement, and no array element replacement.

**Comment by ljestrada** [comment]
> If the former, then I'd say it's worth consideration as a bug.

I just ran into this problem and I'm trying to get past the former, mentioned in your post. It would be nice if support for total replacement of all the array elements is supported or the lack of support is documented--along with the workarounds.

**Comment by markstos** [comment]
@ljestrada We do document that arrays are merged by replacement:

https://github.com/lorenwest/node-config/wiki/Configuration-Files#arrays-are-merged-by-replacement

**Comment by markstos** [comment]
@ljestrada You are welcome to improve the wiki.

**Comment by ljestrada** [comment]
> @ljestrada You are welcome to improve the wiki.

Thx! Completely missed it :-).

I may get away with one of the suggested alternatives.  In essence, I have a bunch of endpoints with a certificate path.  Since the endpoints need to support the same certificate, I'm refactoring the cert path to its own field.  Something like:
"xxx": {
  "certPath": "CERT_PATH", <-- Moved out from each endpoint.
  "endpoints": [
      { "certPath": "CERT_PATH"}, ...
  ]
}

If endpoints needed to support a different certificate, it would be nice to have the "replace the whole array approach:

"xxx": {
  "endpoints": [
      { "certPath": "CERT_PATH1"}, { "certPath": "CERT_PATH2"}, ...
  ]
}

But what I have now will work. Thx again for a prompt reply.

**Comment by ashishblood** [comment]
Hello @markstos and @lorenwest, we required a case where we needed an array substitution for `custom-environment-variables.json` and we implemented it this way. - [final changes](https://github.com/Firstbloodio/node-config/blob/7a58d64c946cc664a81cf6049ad65ce627e18b78/lib/config.js#L1067-L1110)

This might be an incorrect example since it has some other things to replace the variables or even not so performant code.

Something similar can be implemented here?

**Comment by markstos** [comment]
@ashishblood Did you also implement tests which highlight how the changed behavior works? Did check out #569 ? It proposes addressing mapping environment variables using a templating engine. It includes working code on a branch.

**Comment by ashishblood** [comment]
Yes, I did write some tests to make sure that the existing features are intact - [array substitute tests](https://github.com/Firstbloodio/node-config/blob/master/test/15-redact-if-passed-test.js).

The current code skips the substitution of whole array if *any* of the `ENV` variable substitution fails and overrides from environment file in which it is called from.

I just saw the proposal, it looks cool but I wouldn't want to write middlewares/code for a trivial operation of replacing `ENV` variables but that is my opinion. Also, is it this [branch](https://github.com/iMoses/node-config/commits/master) you are trying to pointing me to?

**Comment by lorenwest** [comment]
I agree with @ashishblood that this seems to be more the rounding out of (completing) existing functionality than a middleware function.

**Comment by ashishblood** [comment]
Can I create a PR for this feature to be included with the core lib? 
It'll be really helpful if this lib supports it out of the box, the existing functionality will not be affected by this addition.

**Comment by lorenwest** [comment]
Yes

**Comment by nolazybits** [comment]
having the same requirement.
I need to substitute property in an object in an array.

**Comment by jdmarshall** [comment]
If you have tricky requirements that require imperative logic to work right, node-config does support .js files in the configuration dir.

However I would strongly recommend you make a rule on your project that you only use string interpolation and boolean operations. Don't let people write functions or nobody will be able to read the config files anymore.

```
    "host": process.env.SMTP_HOST ?? "localhost",
    "port": process.env.SMTP_PORT ?? 22`,
    "foo": [SOME_VALUE, "fallback1", "fallback2"]
```

**Comment by markstos** [comment]
Closing due to lack of activity since 2019. Open a new issue if you still have a related problem in 2025. Please include a failing test to illustrate it.
